dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/security/TwoFactorController.java

```diff
@@ -74,142 +70,125 @@
+   */
   @OpenApi.Response(byte[].class)
-  @GetMapping(value = "/qrCode", produces = APPLICATION_OCTET_STREAM_VALUE)
+  @GetMapping(
+      value = {"/showQRCodeAsImage"},
+      produces = APPLICATION_OCTET_STREAM_VALUE)
   @ResponseStatus(HttpStatus.ACCEPTED)
-  public void generateQRCode(
-      @CurrentUser User currentUser, HttpServletResponse response, SystemSettings settings)
-      throws IOException, WebMessageException {
-    if (currentUser == null) {
-      throw new WebMessageException(conflict(ErrorCode.E3027.getMessage(), ErrorCode.E3027));
-    }
-
-    if (currentUser.isTwoFactorEnabled()
-        && !UserService.hasTwoFactorSecretForApproval(currentUser)) {
-      throw new WebMessageException(conflict(ErrorCode.E3022.getMessage(), ErrorCode.E3022));
-    }
-
-    defaultUserService.generateTwoFactorOtpSecretForApproval(currentUser);
-
-    String appName = settings.getApplicationTitle();
-
-    List<ErrorCode> errorCodes = new ArrayList<>();
-
-    String qrContent =
-        TwoFactoryAuthenticationUtils.generateQrContent(appName, currentUser, errorCodes::add);
-
-    if (!errorCodes.isEmpty()) {
-      throw new WebMessageException(conflict(errorCodes.get(0).getMessage(), errorCodes.get(0)));
-    }
+  public void showQRCodeAsImage(@CurrentUser User currentUser, HttpServletResponse response)
+      throws IOException, ConflictException {
+    byte[] qrCode = twoFactorAuthService.generateQRCode(currentUser);
+    response.getOutputStream().write(qrCode);
+  }
 
-    byte[] qrCode =
-        TwoFactoryAuthenticationUtils.generateQRCode(qrContent, 200, 200, errorCodes::add);
+  /**
+   * Shows the generated QR code for the user to scan as a JSON object. Where the QR code (PNG
+   * image) is represented as a base64 encoded string. And the secret encoded in Base32.
+   *
+   * @throws ConflictException
+   */
+  @GetMapping(
+      value = {"/showQRCodeAsJson"},
+      produces = APPLICATION_JSON_VALUE)
+  @ResponseStatus(HttpStatus.OK)
+  public QRCode showQRCodeAsText(@CurrentUser User currentUser) throws ConflictException {
+    byte[] qrCode = twoFactorAuthService.generateQRCode(currentUser);
+    return new QRCode(currentUser.getSecret(), Base64.getEncoder().encodeToString(qrCode));
+  }
 
-    if (!errorCodes.isEmpty()) {
-      throw new WebMessageException(conflict(errorCodes.get(0).getMessage(), errorCodes.get(0)));
-    }
+  public record QRCode(@JsonProperty String base32Secret, @JsonProperty String base64QRImage) {}
 
-    OutputStream outputStream = response.getOutputStream();
-    outputStream.write(qrCode);
+  /**
+   * Enrolls the user in TOTP 2FA and generates a QR code for the user to scan.
+   *
+   * @throws IOException The QR code could not be generated.
+   * @throws ConflictException The user is already enrolled in 2FA.
+   */
+  @OpenApi.Response(byte[].class)
+  @GetMapping(
+      value = {"/qrCode"},
+      produces = APPLICATION_OCTET_STREAM_VALUE)
+  @ResponseStatus(HttpStatus.ACCEPTED)
```

Comment:
These were my initial thoughts below but I think you just moved the endpoint here and I guess we have to keep it like this for backwards compatibility?

> `ACCEPTED` seems an odd choice for a status as the endpoints task seems to be to respond with a QR code which it always does so there is nothing left or unfinished that would make `ACCEPTED` the better choice or am I missing something? In terms of 2FA the process is not complete but the work of this endpoint is complete.
> Since we call `enrollTOTP2FA` that to me looks like a modification of sorts so `GET` seems the wrong HTTP method. Isn't this more of a `PUT` or `POST`?